File: plugins/harness/skills/harness-install/assets/maven/harness-maven-plugin/src/main/java/ai/harness/maven/rules/RequireDocHeadingsRule.java
package ai.harness.maven.rules;

import ai.harness.maven.Finding;

import tools.jackson.databind.JsonNode;
import org.apache.maven.plugin.MojoExecutionException;
import java.nio.file.Path;
import java.util.Collection;
import java.util.stream.Stream;

/**
 * Rule that requires specific headings in documentation files.
 */
public enum RequireDocHeadingsRule implements HarnessCheckRule {
    INSTANCE;
    private static final String CATEGORY = "requireDocHeadings";

    @Override
    public boolean applies(JsonNode manifest) {
        return HarnessCheckHelper.applies(manifest, CATEGORY);
    }

    @Override
    public Collection<Finding> validate(Path root, JsonNode manifest) throws MojoExecutionException {
        final JsonNode catNode = manifest.get(CATEGORY);
        final List<String> headings = HarnessCheckHelper.extractPaths(catNode.get("parameters").get("headings"));
        final JsonNode sourceFilter = catNode.get("parameters").get("sourceFilter");
        final String prefix = sourceFilter.get("prefix").asText();
        final String suffix = sourceFilter.get("suffix").asText();
        final String severity = HarnessCheckHelper.getSeverity(manifest, CATEGORY);
        return HarnessCheckHelper.extractPaths(manifest.get("requireFilesExist").get("parameters").get("paths")).stream()
                .filter(f -> f.startsWith(prefix) && f.endsWith(suffix))
                .flatMap(f -> validateHeadings(root, f, headings, severity).stream())
                .toList();
    }

    private List<Finding> validateHeadings(Path root, String file, List<String> headings, String severity) throws MojoExecutionException {
        final Path filePath = root.resolve(file);
        if (HarnessCheckHelper.isSafeRegularFile(root, filePath)) {
            final String text = HarnessCheckHelper.readFile(root, filePath);
            return headings.stream()
                    .filter(h -> !text.contains(h))
                    .map(h -> new Finding(severity, CATEGORY, "doc missing " + h + ": " + file))
                    .toList();
        }
        return List.of();
    }
}

File: plugins/harness/skills/harness-install/assets/gradle/buildSrc/src/main/kotlin/ai/harness/gradle/rules/RequireDocHeadingsRule.kt
package ai.harness.gradle.rules

import ai.harness.gradle.Finding
import ai.harness.gradle.HarnessCheck
import ai.harness.gradle.HarnessPsiResults
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.contentOrNull
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import java.nio.file.Path

/**
 * Rule that requires documentation files to contain specified headings.
 */
object RequireDocHeadingsRule : HarnessCheckRule {
    override fun applies(manifest: JsonObject): Boolean {
        val category = "requireDocHeadings"
        val catObj = manifest[category]?.jsonObject ?: return false
        val enabled = catObj["enabled"]?.jsonPrimitive?.contentOrNull?.toBoolean() ?: true
        return enabled
    }

    override fun validate(
        manifest: JsonObject,
        root: Path,
        psiResults: HarnessPsiResults?,
    ): Collection<Finding> {
        val category = "requireDocHeadings"
        val catObj = manifest[category]?.jsonObject
        val parametersObj = catObj?.get("parameters")?.jsonObject
        return if (catObj == null || parametersObj == null) {
            emptyList()
        } else {
            val sourceCategory = HarnessCheck.stringFrom(parametersObj, "sourceFilesFromCategory")
            val sourceFiles =
                HarnessCheck.stringArrayFrom(
                    manifest[sourceCategory]?.jsonObject?.get("parameters")?.jsonObject,
                    "paths",
                )
            val sourceFilterObj = parametersObj["sourceFilter"]?.jsonObject
            val prefix = HarnessCheck.stringFrom(sourceFilterObj, "prefix")
            val suffix = HarnessCheck.stringFrom(sourceFilterObj, "suffix")
            val headings = HarnessCheck.stringArrayFrom(parametersObj, "headings")
            sourceFiles.filter { sourceFile -> sourceFile.startsWith(prefix) && sourceFile.endsWith(suffix) }.flatMap { docPath ->
                headings
                    .filter { heading ->
                        !HarnessCheck.readSafe(root, docPath).contains(heading)
                    }.map { heading ->
                        Finding(HarnessCheck.severityOf(manifest, category), category, "doc missing $heading: $docPath")
                    }
            }
        }
    }
}
